### PR TITLE
add skeleton for updateManualVol cron job

### PR DESCRIPTION
**Pull request by chuddster**
Tickets: 

1. [ManualVolOracle update script](https://www.notion.so/ribbonfinance/ManualVolOracle-update-script-a5ff11b820754fcf9f0e50059c7db848)
2. [Notification tooling to give visibility to week's selected strike price](https://www.notion.so/ribbonfinance/Notification-tooling-to-give-visibility-to-week-s-selected-strike-price-0c20480e586a4aa79de0cdd951d3af88)
3. [Wire up master cron to mainnet addresses](https://www.notion.so/ribbonfinance/Wire-up-master-cron-to-mainnet-addresses-bf530e1206cf4978b6a5ef831b04979b)

### CODE REVIEW

File: scripts/masterCron.ts [hunk 1]
```diff
@@ -37,6 +38,7 @@ const signer = getDefaultSigner("m/44'/60'/0'/0/0", network).connect(provider);
 
 let gasLimits = {
   volOracleCommit: 85000,
+  volOracleAnnualizedVol: 0,
   settleAuction: 0,
```
Comment (chuddster) [inline scripts/masterCron.ts:42]: wdyt these gas limits should be

File: constants/constants.ts [hunk 2]
```diff
@@ -75,6 +75,7 @@ export const BYTES_ZERO =
  */
 
 export const VOL_ORACLE = "0x4df938e57fD4Ad1dFDdDEEb1B4cFAbAB19E33A0E";
+export const MANUAL_VOL_ORACLE = "0xe24854968CDbCfa049ddd6F089b5C552D181eCC4";
```
Comment (chuddster) [inline constants/constants.ts:79]: add strikeselection/optionspremiumpricer addresses alongside the vaults

File: constants/deployments-mainnet-cron.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        "strikeSelection": "",
+        "optionsPremiumPricer": "0x947398a2C3f5dD6983bb935FEBFcF076C870e23e"
+      },
+      "RibbonThetaVaultETHPut": {
+        "address": "0xEC1c50724CF7a618C6cdA6CFEa5c9064AFc98E84",
+        "strikeSelection": "0xaAA24915AfFab8A83af74D5A1472F4fE7AD08D71",
+        "optionsPremiumPricer": "0xE922806E610A6bA9A828a3b2Ea6A9936182FFC7D"
+      }
+    },
+    "gnosisLibrary": "",
```
Comment (chuddster) [inline constants/deployments-mainnet-cron.json:20]: need address for kovan + mainnet gnosis library

File: constants/deployments-mainnet-cron.json [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    "vaultLifecycle": "0x9bCF7E35E8B852de70ddB993BC461d1f4624e5AC",
+    "univ3pools": {
+      "wethusdc": "",
+      "wbtcusdc": ""
+    }
+  },
+
```
Comment (chuddster) [inline constants/deployments-mainnet-cron.json:27]: prefer this custom heirarchical json to read from in master cron. will need to update addresses manually when we deploy new vaults

File: scripts/masterCron.ts [hunk 5]
```diff
@@ -58,6 +63,110 @@ const getTopOfPeriod = async (provider: any, period: number) => {
   return topOfPeriod;
 };
 
+async function getStrikePrice(
```
Comment (chuddster) [inline scripts/masterCron.ts:66]: copying from logic in vault lifecycle [commitAndClose](https://github.com/ribbon-finance/ribbon-v2/blob/bd9b69f2b68132cea8d418709c00acab57e815e0/contracts/libraries/VaultLifecycle.sol#L53)

File: scripts/masterCron.ts [hunk 6]
```diff
@@ -58,6 +63,110 @@ const getTopOfPeriod = async (provider: any, period: number) => {
...
+  }
+
+  let isPut = (await vault.vaultParams()).isPut;
+
+  let strike = await strikeSelection.getStrikePrice(expiry, isPut);
+
+  return expiry, isPut, strike;
+}
+
+async function getOptionPremium(
```
Comment (chuddster) [inline scripts/masterCron.ts:97]: copying from logic in vault lifecycle [commitAndClose](https://github.com/ribbon-finance/ribbon-v2/blob/bd9b69f2b68132cea8d418709c00acab57e815e0/contracts/libraries/VaultLifecycle.sol#L75)

File: scripts/masterCron.ts [hunk 7]
```diff
@@ -58,6 +63,110 @@ const getTopOfPeriod = async (provider: any, period: number) => {
...
+    ).underlying,
+    (
+      await vault.vaultParams()
+    ).asset,
+    strikePrice,
+    expiry,
+    isPut
+  );
+
+  let premium = await gnosisAuction.getOTokenPremium(
```
Comment (chuddster) [inline scripts/masterCron.ts:132]: TODO: adjust premium based on yvusdc -> usdc conversion rate steth -> eth conversion rates

File: scripts/masterCron.ts [hunk 8]
```diff
@@ -8,15 +8,19 @@ import {
   getDefaultSigner,
 } from "./helpers/getDefaultEthersProvider";
 import moment from "moment";
-import deployments from "../constants/deployments.json";
+import deployments from "../constants/deployments-mainnet-cron.json";
```
Comment (kenchangh) [inline scripts/masterCron.ts:11]: Is there a reason why we have to use `deployments-mainnet-cron.json` instead? We're generating `constants/deployments.json` using the `yarn export-deployments` command so we have a single source of truth

File: constants/deployments-mainnet-cron.json (hunk 3 above)
Comment (kenchangh) [reply constants/deployments-mainnet-cron.json:20]: kovan - `0x3D1D40101E185E1fE3aedCF05FE2F5BA0Af1f25D`
mainnet - `0x0b7fFc1f4AD541A4Ed16b40D8c37f0929158D101`

File: scripts/masterCron.ts (hunk 1 above)
Comment (kenchangh) [reply scripts/masterCron.ts:46]: 50k gas

File: scripts/masterCron.ts [hunk 9]
```diff
@@ -58,6 +64,115 @@ const getTopOfPeriod = async (provider: any, period: number) => {
...
+    method: "public/get_volatility_index_data",
+    params: {
+      currency: underlying,
+      // resolution is in minutes, we multiply by 60 to get seconds
+      start_timestamp: (latestTimestamp - resolution * 60).toString(),
+      end_timestamp: latestTimestamp.toString(),
+      resolution: resolution,
+    },
+  };
+  var ws = new WebSocket("wss://www.deribit.com/ws/api/v2");
```
Comment (kargakis) [inline scripts/masterCron.ts:162]: Are you missing importing WebSocket from socket.io? Also shouldn't socket.io be tracked in package.json?

File: scripts/masterCron.ts [hunk 10]
```diff
@@ -58,6 +64,115 @@ const getTopOfPeriod = async (provider: any, period: number) => {
...
+    await vault.premiumDiscount()
+  );
+
+  return premium;
+}
+
+async function getAnnualizedVol(underlying: string, resolution: string) {
+  const latestTimestamp = (await provider.getBlock("latest")).timestamp;
+
+  var msg = {
```
Comment (kargakis) [inline scripts/masterCron.ts:150]: nit: const?

**Comment by kargakis** [comment]
Wondering whether some of these cronjobs can be moved into a decentralized execution network like Gelato or Keeper. Have you folks spent anytime thinking about this or is it not a concern at this point?

File: constants/deployments-mainnet-cron.json (hunk 3 above)
Comment (chuddster) [reply constants/deployments-mainnet-cron.json:20]: I actually meant the gnosis library, but forgot that it wasn't external :)

File: scripts/masterCron.ts (hunk 9 above)
Comment (chuddster) [reply scripts/masterCron.ts:162]: don't think we need imports

**Comment by chuddster** [comment]
> Wondering whether some of these cronjobs can be moved into a decentralized execution network like Gelato or Keeper. Have you folks spent anytime thinking about this or is it not a concern at this point?

I actually only came across gelato only a few days ago - after we got most of this set up. at this point we are prioritizing getting v2 out the door and running smoothly. but maybe in the future we will take it from the backlog

File: scripts/masterCron.ts (hunk 1 above)
Comment (chuddster) [reply scripts/masterCron.ts]: changed gas limits to all lmk if reasonable

File: scripts/masterCron.ts [hunk 11]
```diff
@@ -126,7 +285,7 @@ async function runTX(
     // someone already called new weeks rollToNextOption
     if (
       method === "rollToNextOption" &&
-      (await vault.currentOption()) === constants.AddressZero
+      (await vault.currentOption()) != constants.AddressZero
```
Comment (kenchangh) [inline scripts/masterCron.ts:288]: nit use `!==`

File: scripts/masterCron.ts [hunk 12]
```diff
@@ -67,35 +219,42 @@ async function settleAuctionsAndClaim(
...
-      ]);
+      let newGasPrice = (await gas(network)).toString();
+
+      const tx = await gnosisAuction
+        .connect(signer)
+        .claimFromParticipantOrder(
+          auctionID,
+          [encodeOrder(await vault.auctionSellOrder())],
+          { gasPrice: newGasPrice, gasLimit: gasLimits["claimAuctionOtokens"] }
+        );
```
Comment (kenchangh) [inline scripts/masterCron.ts:257]: We can remove this for Theta Vaults, this only applies for delta vaults

**Review by kenchangh — APPROVED** [review]

**Pull request merged**